pandas/core/frame.py

```diff
@@ -4317,7 +4321,7 @@ def set_index(
                     f"received array of length {len(arrays[-1])}"
                 )
 
-        index = ensure_index_from_sequences(arrays, names)
+        index = ensure_index_from_sequences(arrays, names, current_dtype)
```

Comment:
I think you *might* able to make a much more direct change.

line 4311, something like
```
col = frame[col]
arrays.append(Index(col, dtype=col.dtype)
names.append(col)
```

Reply:
and you don't need the rest of the changes

Reply:
yes good point, I'll work on this

Reply:
@jreback I tested this out by directly appending the Index like in your example, and it did not override the dtype. This looks to be because when the call stack gets inside `ensure_index_from_sequences` where it actually creates the index on the line `return Index(sequences[0], name=names)`, if you do it this way you essentially get `return Index(Index(col, col.dtype), name=names)`. In a sense, the inner constructor in this example has your custom dtype, but by not setting that param in the outer constructor, it just reverts to type inference.

I think we need changes more like what I initially attempted e.g. we need to directly set the dtype in the Index constructor inside of `ensure_index_from_sequences`.